spec.bs

```diff
@@ -1112,19 +1112,148 @@ For now, see [https://github.com/wicg/observable#operators](https://github.com/w
+          reason=].
+
+    1. Let |observer| be a new [=internal observer=], initialized as follows:
+
+       : [=internal observer/next steps=]
+       :: 1. [=Invoke=] |predicate| with the passed in <var ignore>value</var>, and let |passed| be
+             the returned value.
+
+             If <a spec=webidl lt="an exception was thrown">an exception |E| was thrown</a>, then
+             [=reject=] |p| with |E|, and [=AbortController/signal abort=] |controller| with |E|.
+
+         1. If |passed| is false, then [=resolve=] |p| with false, and [=AbortController/signal
+            abort=] |controller|.
+
+       : [=internal observer/error steps=]
+       :: [=Reject=] |p| with the passed in <var ignore>error</var>.
+
+       : [=internal observer/complete steps=]
+       :: [=Resolve=] |p| with true.
+
+    1. <a for=Observable lt="subscribe to an Observable">Subscribe</a> to [=this=] given |observer|
+       and |internal options|.
+
+    1. Return |p|.
 </div>
 
 <div algorithm>
   The <dfn for=Observable method><code>find(|predicate|, |options|)</code></dfn> method steps are:
 
-    1. <span class=XXX>TODO: Spec this and use |predicate| and |options|.</span>
+    1. Let |p| [=a new promise=].
+
+    1. Let |controller| be a [=new=] {{AbortController}}.
+
+    1. Let |internal options| be a new {{SubscribeOptions}} whose {{SubscribeOptions/signal}} is the
+       result of [=creating a dependent abort signal=] from the list
+       «|controller|'s [=AbortController/signal=], |options|'s
+       {{SubscribeOptions/signal}} if non-null», using {{AbortSignal}}, and the [=current realm=].
+
+    1. If |internal options|'s {{SubscribeOptions/signal}} is [=AbortSignal/aborted=], then:
+
+       1. [=Reject=] |p| with |internal options|'s {{SubscribeOptions/signal}}'s
+          [=AbortSignal/abort reason=].
+
+       1. Return |p|.
+
+    1. [=AbortSignal/add|Add the following abort algorithm=] to |internal options|'s
+       {{SubscribeOptions/signal}}:
+
+       1. [=Reject=] |p| with |internal options|'s {{SubscribeOptions/signal}}'s [=AbortSignal/abort
+          reason=].
+
+    1. Let |observer| be a new [=internal observer=], initialized as follows:
+
+       : [=internal observer/next steps=]
+       :: 1. [=Invoke=] |predicate| with the passed in |value|, and let |passed| be the returned
+             value.
+
+             If <a spec=webidl lt="an exception was thrown">an exception |E| was thrown</a>, then
+             [=reject=] |p| with |E|, and [=AbortController/signal abort=] |controller| with |E|.
+
+         1. If |passed| is true, then [=resolve=] |p| with |value|, and [=AbortController/signal
+            abort=] |controller|.
+
+       : [=internal observer/error steps=]
+       :: [=Reject=] |p| with the passed in <var ignore>error</var>.
+
+       : [=internal observer/complete steps=]
+       :: [=Resolve=] |p| with false.
```

Comment:
This should reject (With a `RangeError` maybe, to follow suit with `first` and `last`) because the value was not found. What if the consumer is trying to find `false`? `observable.find(v => v === false)`? 

In fact, there should probably be a default option to override that (as there probably should be for `first` and `last`): `observable.find(predicate, { defaultValue: 'Did not find the value' })`.

Reply:
The default behavior should really be to match arrays / iterators, which return `undefined` in the case that a match was not found.

(Which, yes, makes it hard to search for `undefined`. But I don't think this should break with precedent)

Reply:
Done! Hmm, in that case I'm thinking `first()` and `last()` should probably do the same then, instead of using `RangeError`s, even if [Iterator helpers is unlikely to get those operators](https://github.com/WICG/observable/issues/132#issuecomment-2038410116).

Reply:
(I've left a comment about this on https://github.com/WICG/observable/pull/131#issuecomment-2083248807 for `first()` and `last()`.)

Reply:
**TL;DR: I can concede this if we think it makes everything more palatable for the committee. I do strongly advise we allow a `defaultValue` configuration though, to cover the corner case**
 
@domfarolino @bakkot FWIW... the behavior of RxJS 1 - 6's `Observable.prototype.toPromise()` for about 10 years was to emit `undefined` in cases where an observable completed without emitting a value.

Over the years, this caused enough issues and small bugs that it was agreed that `firstValueFrom(observable)` and `lastValueFrom(observable)` was to throw an `EmptyError` (specific to RxJS), unless a `defaultValue` was specified.

`toPromise()`, `firstValueFrom`, and `lastValueFrom` are _heavily used_ as a means of interop with promise-based APIs and async/await.

On the other hand, RxJS's `find` operator, does indeed emit `undefined` if it can't find what you're looking for. This has largely been okay and not caused issues, however, `find` is _rarely_ used from what I've seen over the years. We do have a `first(predicate, defaultValue?)` operator that will throw an `EmptyError` if it doesn't find the value, and no default was provided.

Now... all of this said, the major issue with RxJS's `toPromise()` from what I'm seeing through all filed issues, was the fact that (since RxJS 5 was originally written against TypeScript 1.8) the return type of `toPromise()` was wrong, being `Promise<T>` instead of `Promise<T | undefined>`, and the name was too general, as it always gave you the last value of the observable. And given that people largely wanted `Promise<T>`, and we also wanted to remove the method from `Observable`, we changed to have it reject with an `EmptyError` and named it more appropriately, as `lastValueFrom()`. Then we added `firstValueFrom()` because it was common pattern for people to do `observable.take(1).toPromise()`. 

I'm totally okay with moving back to having it return `undefined` if the item isn't found, or nothing is emitted, if we think that helps move things along. But I still recommend a configurable default value. I'm not sure if there's prior art for that though.

Reply:
Also: Apologies, I merged #133 before I saw this thread.

Reply:
I'm not as attached to returning `undefined` for `first` and `last`, though it does seem more consistent/usable than an error would be. Mostly I want `find` specifically to match arrays/iterators, since it exists under that name already, whereas `first` and `last` don't.

I agree that in theory a configurable fallback value for `find` (and the other two) would be nice, though in practice I can't think of a time that's ever actually come up for me.

Reply:
Yeah +1 to pretty much all of that. I'll work on making #131 #144 not use `RangeError`, instead yielding `undefined` in the vacuous case, and then we have https://github.com/WICG/observable/issues/132 still open which I think is important but not a hard blocker for initially launching.